dlib/image_transforms/draw.h

```diff
@@ -561,17 +571,75 @@ namespace dlib
                     rgb_alpha_pixel temp = alpha_pixel;
                     temp.alpha = static_cast<unsigned char>((right_boundary[i-1]-x)/delta*max_alpha);
                     if (valid_area.contains(p))
-                        assign_pixel(image(p.y(), p.x()),temp);
+                        assign_pixel(img[p.y()][p.x()],temp);
                 }
             }
         }
     }
 
-    template <typename image_type>
-    inline void draw_solid_convex_polygon (
+// ---------------------------------------------------------------------------------------
+
+    template <typename image_type, typename pixel_type>
+    void draw_solid_polygon (
         image_type& image,
-        const polygon& poly
-    ) { fill_convex_polygon(image, poly, 0); }
+        const polygon& poly,
+        const pixel_type& color,
+        const rectangle& area = rectangle(std::numeric_limits<long>::min(), std::numeric_limits<long>::min(),
+                                          std::numeric_limits<long>::max(), std::numeric_limits<long>::max())
+    )
+    {
+        image_view<image_type> img(image);
+        const rectangle valid_area(get_rect(image).intersect(area));
+        const rectangle bounding_box = poly.get_rect();
+        if (bounding_box.intersect(valid_area).is_empty())
+            return;
+
+        const long left = std::max(valid_area.left(), bounding_box.left() + 1);
+        const long top = std::max(valid_area.top(), bounding_box.top() + 1);
+        const long right = std::min(valid_area.right(), bounding_box.right() - 1);
+        const long bottom = std::min(valid_area.bottom(), bounding_box.bottom() - 1);
```

Comment:
Same thing here.  Totally use the rectangle methods to do this kind of stuff, it's way easier :)

Like this is `shrink_rect(valid_area.intersect(bounding_box),1)` right?